### PR TITLE
RDP Redirect Microphones

### PR DESCRIPTION
**Pull request by countchappy**
Adding the option to redirect microphones from local computer

### CODE REVIEW

**Comment by countchappy** [comment]
@sparerd and @kmscode I was able to test this and confirm it showed up correctly in mRemoteNG's configuration window. However, I was unable to connect to my terminal server and confirm whether or not the microphones are redirecting (windows audio service is disabled) 😞 . If one of you could check this it would be greatly appreciated.

**Comment by kmscode** [comment]
@countchappy - please update the ConfVersionSerialized test to check for version 2.8. This will prevent the jenkins build failure.

I'll see about getting this tested early this week (either later today, or at some point tomorrow).

**Comment by countchappy** [comment]
@kmscode Pushed a commit. But it's still getting a failed Jenkins test ... I've never used Jenkins so I do not know how to check why it is failing.

**Comment by countchappy** [comment]
@kmscode Okay I think I got the issue squared away 😆

**Comment by sparerd** [comment]
@countchappy Sorry, the failing test was my fault 🙁.

It was a brittle test to begin with, so I've fixed it for good now (it was making another PR fail for no good reason too).

**Comment by sparerd** [comment]
Hey @countchappy ! Just wanted to check in and let you know we haven't forgotten about your PR. Thanks again for submitting it!

**Comment by countchappy** [comment]
@sparerd no worries. I was wondering why I was getting emails at 11:00 at night 😆

**Comment by rheingold** [comment]
If I may comment: I hjave just added the new texts from language resource into the Czech loc. however when I studied your code and compared it to the documentation by MSoft, then - if Im not mistaking - this redirect does not redirect the microphones only (filter on the mmedia input source), but all currently available audio inputs, depending which one is selected as th eaudio source (typically the default one). So the lables should not be "Microphone" but rather "Audio input" (which i have used within the translation) or - as to match MSoft vocab - rather "Remote audio recording".

**Comment by countchappy** [comment]
@rheingold Thanks for pointing that out to me. I will update and commit 😄

**Comment by countchappy** [comment]
@rheingold Just pushed the commit for that change. The variables are still named according to `RedirectMicrophones`. However, in a sense of continuity I will get those switched over.

**Comment by rheingold** [comment]
@countchappy: Ok, nice :-) Thanks for the backward compatibility of the resource var names, I hoped I wont have to update my commit/push again ;-)

**Comment by kmscode** [comment]
@countchappy
The good news: I tested this. 
The bad news: Windows on the remote host didn't detect my USB microphone or my builtin microphone. 
The good/bad news: Neither did mstsc ("good" as in I don't think your code is broken)....

**Comment by kmscode** [comment]
OK... So I tested "update 4" since as I was writing my comment "update 5" was committed.

File: mRemoteV1/Connection/Protocol/RDP/Connection.Protocol.RDP.cs [hunk 1]
```diff
@@ -547,6 +547,7 @@ private void SetRedirection()
 				_rdpClient.AdvancedSettings2.RedirectPrinters = _connectionInfo.RedirectPrinters;
 				_rdpClient.AdvancedSettings2.RedirectSmartCards = _connectionInfo.RedirectSmartCards;
 				_rdpClient.SecuredSettings2.AudioRedirectionMode = (int)_connectionInfo.RedirectSound;
+                _rdpClient.AdvancedSettings8.AudioCaptureRedirectionMode = ((int)_connectionInfo.RedirectAudioInput) != 0;
```
Comment (kmscode) [inline mRemoteV1/Connection/Protocol/RDP/Connection.Protocol.RDP.cs]: _rdpClient.AdvancedSettings8.AudioCaptureRedirectionMode = Convert.ToBoolean(_connectionInfo.RedirectMicrophones);

Looks a little cleaner... But at a minimum you don't need the cast or parenthesis.

**Comment by kmscode** [comment]
And I have no idea why Jenkins is failing to build after the latest update...

**Comment by sparerd** [comment]
From the build log:
> Connection\ConnectionInfo.cs(292,124): error CS1061: 'Settings' does not contain a definition for 'ConDefaultRedirectAudioInput' and no extension method 'ConDefaultRedirectAudioInput' accepting a first argument of type 'Settings' could be found (are you missing a using directive or an assembly reference?)

**Comment by countchappy** [comment]
@kmscode and @sparerd - Yeah, Will get that fixed and updated

**Comment by kmscode** [comment]
I saw that, but it looks like it was there... But it's also been a long day... I'm done... 😖

**Comment by countchappy** [comment]
@kmscode & @sparerd Okay, I think I fixed my boo boo haha

**Comment by countchappy** [comment]
@kmscode I updated the code per your review _(removed the cast and parenthesis)_. In regards to the redirect not working, I will do some digging to find out why. I can't do it right at the moment, but should be able to sometime later tonight.

**Comment by countchappy** [comment]
@kmscode Hey, sorry for such a delay. Unfortunately, I have been unable to find the answer to this. I will post it to stackoverflow and try my luck there.

**Comment by kmscode** [comment]
No problem... 

My only other thought (without having looked into it/testing it further at this point) - does Windows Server 2012 R2 (or any server version) accept microphone input even with mstsc? I'm provisioning a Windows 10 VM that I'll try to test with...

**Comment by bearlikelion** [comment]
As long as Windows audio is enabled on the server and local device redirection is enabled on AD/local registry.

Not enabling microphone redirection was a deal breaking for me working (mic is passed through to VM).
So I've switched to VMware Workstation temporarily.

**Comment by 4everwo** [comment]
Hi, is this enhancement still under development? I see last merge failed but no details on what happened in the end.
Thanks

**Comment by countchappy** [comment]
Hello Charles,

I am no longer working on this feature. I had (at the time) implemented the
feature per how MSDN said it should be, but we couldn't get it to work when
testing. Eventually, I had other things come up that took time away from
what little time I had. Sorry for any inconvenience this may have caused
you.

Regards,
David

**Comment by kmscode** [comment]
I didn't see any other Issues/Requests for this functionality, so closing this PR as Stale since it's been a while. Feel free to reopen/open a new issue/enhancement request accordingly.

**Comment by allencodes** [comment]
Looks like this one was semi-abandoned. :(
As of v1.76.20.24615 the redirection of audio recording is not available. Just checked and it Is working fine using native RDP client. 
With the Microsoft push to using Teams (instead of Skype), this is coming to the front burner for me again. Any chance of getting this request re-engaged?

**Comment by countchappy** [comment]
It has been a while since I have done anything on this. I will dust off the keyboard and see if I can work some magic.

**Comment by etherjack** [comment]
I just discovered there was a new preview version released since I first installed (1.77). I see the option for capture incoming audio and my RDP session lists Input as 'Remote Audio'. I have not actually tested the functionality yet, but at least the options are there :)

**Comment by mar10** [comment]
I also use Teams/Skype and using a microphone is essential.
Tried the audio captutr option in 1.77.1.27654, but it doesn't work (the remote windows is not offering a microphone). 
It works with RDP though.

**Comment by nunnsby** [comment]
Hi there, any updates on his issue of "Mic-In" / Sending Audio to the remote session via mRemoteNG?

**Comment by countchappy** [comment]
@nunnsby I setup a VM, I should be able to do some testing this weekend. Will keep you posted.

**Comment by etherjack** [comment]
As of v1.77.0.41252 this is working for me. I run MS Teams on a VM
regularly and the RDP connection via mRemoteNG works fine.

**Comment by maxkoshevoi** [comment]
Hi, is there still no option to pass microphone from local computer?

**Pull request closed without merging**